### PR TITLE
MGMT-5456 docs: hive integ - z-stream support

### PR DESCRIPTION
**Pull request by danielerez**
Added documentation for: 'Hive Integration - Selecting OpenShift Versions'

### CODE REVIEW

**Comment by danielerez** [comment]
@filanov @rollandf @jparrill @flaper87 PTAL

File: docs/kube-api-custom-ocp-versions.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Custom OpenShift Versions (Dynamic z-stream Support)
```
Comment (rollandf) [inline docs/kube-api-custom-ocp-versions.md:1]: Maybe add a reference to this page under:
https://github.com/openshift/assisted-service/blob/master/docs/Kube-API.md#clusterdeployment

File: docs/kube-api-custom-ocp-versions.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+export OPENSHIFT_VERSIONS='{"4.7":{"rhcos_image":"https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.7/4.7.0/rhcos-4.7.0-x86_64-live.x86_64.iso","rhcos_version":"47.83.202102090044-0"}}'
+```
+
+### Deploy ClusterImageSet
+
+Deploy a ClusterImageSet with the requested release image.
+
+E.g.
+```
+apiVersion: hive.openshift.io/v1
```
Comment (rollandf) [inline docs/kube-api-custom-ocp-versions.md:74]: Maybe have the yamls under also in `docs/crds`

File: docs/kube-api-custom-ocp-versions.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  name: openshift-v4.7.0
+spec:
+  releaseImage: quay.io/openshift-release-dev/ocp-release:4.7.0-x86_64
+```
+
+### [OPENSHIFT_VERSIONS](https://github.com/openshift/assisted-service/blob/master/Makefile#L36)
+
+The OPENSHIFT_VERSIONS environment variable is a map of OpenShift version (y-stream) to related properties as release version/image and RHCOS version/image: see [default_ocp_versions.json](https://github.com/openshift/assisted-service/blob/master/default_ocp_versions.json)
+
+Defining this var prior to starting the service would override all versions supported by the service. I.e. by default, [default_ocp_versions.json](https://github.com/openshift/assisted-service/blob/master/default_ocp_versions.json) is used for setting the supported versions. Defining OPENSHIFT_VERSIONS with a similar json format should override these.
```
Comment (filanov) [inline docs/kube-api-custom-ocp-versions.md:32]: need to explain that only minimal data is required. `rhcos_image` `rhcos_version`

File: docs/kube-api-custom-ocp-versions.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+* OPENSHIFT_VERSIONS is defined with a map of "x.y" -> ```rhcos_image``` and ```rhcos_version```.
+* Upon starting the service, the relevant host [boot-files](https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L16) are uploaded to S3/File storage.
+* ClusterImageSet with a new ```releaseImage``` is deployed to Hive cluster.
+* The name of the ClusterImageSet is referenced in the ClusterDeployment (```spec.provisioning.imageSetRef``` property).
+* Upon deploying the ClusterDeployment:
+  * ClusterImageSet is fetched according to ```imageSetRef```.
+  * ```releaseImage``` is read from the ClusterImageSet.
+  * The ```release_version``` is extracted from the release image (using ```oc adm release info```) - in major.minor.patch format.
+  * According to the major.minor as key, the relevant entry is updated in the [versions cache](https://github.com/openshift/assisted-service/blob/master/internal/versions/versions.go). I.e. ```release_version``` and ```display_name``` properties.
+* Finally, a new cluster is created with the newly added [openshift_version](https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L4106).
```
Comment (filanov) [inline docs/kube-api-custom-ocp-versions.md:46]: if I remember correctly there is another way to mark specific lines with blame or something like that, this link will not be relevant if someone adds or removes lines from the swagger.

File: docs/kube-api-custom-ocp-versions.md (hunk 2 above)
Comment (filanov) [reply docs/kube-api-custom-ocp-versions.md:74]: +1

File: docs/kube-api-custom-ocp-versions.md (hunk 1 above)
Comment (mhrivnak) [inline docs/kube-api-custom-ocp-versions.md:1]: nit: I don't think this intends to discuss "custom versions". Red Hat makes releases and decides what the versions are, and here we are just helping the user select versions. I would refer to this as "Selecting OpenShift Versions", "Choosing OpenShift Versions", or similar.

Likewise, the fact that the API is "dynamic" does not mean much to the user. They're using a k8s API, so it's assumed that API use should not require restarting a service.

File: docs/kube-api-custom-ocp-versions.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Custom OpenShift Versions (Dynamic z-stream Support)
+
+As part of [Hive Integration](https://github.com/openshift/assisted-service/blob/master/docs/Kube-API.md), a means to dynamically (without restarting the service) update an OCP version is required.
```
Comment (mhrivnak) [inline docs/kube-api-custom-ocp-versions.md:3]: Here also I don't think it's worth bragging about not requiring a service restart.

File: docs/kube-api-custom-ocp-versions.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+This is mainly useful for disconnected environments that use the kube-api support. E.g. updating the release image to a new z-stream version.
```
Comment (mhrivnak) [inline docs/kube-api-custom-ocp-versions.md:4]: I would tweak this wording to target environments with mirrored releases. Customers mirror content for more reasons than just having disconnected networks.

**Review by mhrivnak — CHANGES_REQUESTED** [review]
Thanks for adding docs! Please coordinate with #1592, which implements the k8s API for setting `OPENSHIFT_VERSIONS`.

I think we probably don't want to get into such low-level detail when it comes to documenting the k8s API. `OPENSHIFT_VERSIONS` is an implementation detail that the user won't interact with in that case. It may be more productive to define this environment variable's behavior, which is complex as you illustrate, in documentation that is focused on all the environment variables and their semantics.

Separately, it is great to have an end-to-end walk-through of how to use the k8s APIs, including those from hive, to accomplish mirroring.

**Pull request merged**